js/app/factory/calendarFactory.js

```diff
@@ -0,0 +1,326 @@
+	};
+
+
+	context.webcal = function(props) {
+		const sourceProp = props['{' + DavClient.NS_CALENDARSERVER + '}source'];
+
+		if (Array.isArray(sourceProp)) {
+			const source = sourceProp.find(function(source) {
+				return (DavClient.getNodesFullName(source) === '{' + DavClient.NS_DAV + '}href');
+			});
+
+			return source ? source.textContent : null;
+		} else {
+			return null;
+		}
+	};
+
+	context.calendarSkeleton = function(props, userPrincipal, publicMode) {
+		const simple = {};
+		const currentUser = context.getUserFromUserPrincipal(userPrincipal);
+
+		simple.color = context.color(props);
+		simple.displayname = context.displayname(props);
+		simple.components = context.components(props);
+		simple.order = context.order(props);
+
+		simple.writable = context.acl(props, userPrincipal);
+		simple.owner = context.owner(props);
+		simple.enabled = context.enabled(props, simple.owner, currentUser);
+
+		simple.shares = context.shares(props, simple.owner);
+
+		const [shareable, publishable] = context.shareableAndPublishable(props, simple.writable, publicMode);
+		simple.shareable = shareable;
+		simple.publishable = publishable;
+
+		const [published, publishurl, publicurl] = context.publishedAndPublishURL(props, publicMode);
+		simple.published = published;
+		simple.publishurl = publishurl;
+		simple.publicurl = publicurl;
+
+		// always enabled calendars in public mode
+		if (publicMode) {
+			simple.enabled = true;
+			simple.writable = false;
+		}
+
+		simple.writableProperties = (currentUser === simple.owner) && simple.writable;
+
+		return simple;
+	};
+
+	context.getUserFromUserPrincipal = function(userPrincipal) {
+		if (userPrincipal.endsWith('/')) {
+			userPrincipal = userPrincipal.slice(0, -1);
+		}
+
+		const slashIndex = userPrincipal.lastIndexOf('/');
+		return userPrincipal.substr(slashIndex + 1);
+	};
+
+	/**
+	 * get a calendar object from raw xml data
+	 * @param body
+	 * @param {string} userPrincipal
+	 * @param {boolean} publicMode
+	 * @returns {Calendar}
+	 */
+	this.calendar = function(body, userPrincipal, publicMode=false) {
+		const href = body.href;
+		const props = body.propStat[0].properties;
+
+		const simple = context.calendarSkeleton(props, userPrincipal, publicMode);
+		return Calendar(href, simple);
+	};
+
+	/**
+	 * get a webcal object from raw xml data
+	 * @param body
+	 * @param {string} userPrincipal
+	 * @param {boolean} publicMode
+	 * @returns {WebCal}
+	 */
+	this.webcal = function(body, userPrincipal, publicMode=false) {
+		const href = body.href;
+		const props = body.propStat[0].properties;
+		const currentUser = context.getUserFromUserPrincipal(userPrincipal);
+
+		const simple = context.calendarSkeleton(props, userPrincipal, publicMode);
+		simple.href = context.webcal(props);
+
+		// WebCal is obviously not writable
+		simple.writable = false;
+		simple.writableProperties = (currentUser === simple.owner);
+
+		// WebCal subscriptions are neither publishable nor shareable
```

Comment:
Please just clean the TODO above and look what's about the rgba safari workaround.